wallet: Improve Fee Priority Code for Clarity

Tzadiko opened the pull request:
Hello,

I'm working on https://github.com/monero-project/monero/issues/9799 . After speaking to @nahuhh , I'd like to contribute in two steps. The first step is to clean the existing code, and then addressing the issue. This makes reviewing my changeset easier.

In this changeset there are no changes to functionality. I added fee_priority .h and .cpp. Here I added the enumeration FeePriority to replace the integer values which were confusing to understand when trying to address the original issue linked above.

I will now be looking at addressing the original issue linked, working off this 'clean up' branch.

Tzadiko commented:
It built locally for me on Ubuntu, but it's failing on other distros. I'll take a look after lunch.

Tzadiko commented:
All tests pass locally. As was communicated to me (by @jeffro256 on Matrix), the test that failed here fails sporadically.

0xFFFC0000 commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        static FeePriority FromIntegral(const uint32_t priority)
+        {
```
if (req.priority > 4) {
         LOG ERROR;
         er.message = "Invalid priority value. Must be between 0 and 4.";
     }

Tzadiko replied:
Thanks, please see the commit where I add this: 69aefe5a7f67ce4bb35f9ab4de20e93bfb574033

0xFFFC0000 reviewed (approved):
No clear correctness bugs stand out. The new code is more verbose and also more readable.

vtnerd commented on src/wallet/wallet2.cpp:
```diff
@@ -8632,16 +8638,21 @@ uint64_t wallet2::adjust_mixin(uint64_t mixin)
-      const uint64_t base_fee = get_base_fee(1);
-      const double fee_level = base_fee * (use_per_byte_fee ? 1 : (12/(double)13 / (double)1024));
+      const uint64_t base_fee = get_base_fee(FeePriority::Unimportant);
+      const double fee_level = base_fee * (use_per_byte_fee ? 1 : (12. / 13. / 1024.));
```
The `double` changes seems unrelated and unnecessary (revert).

Tzadiko replied:
Done

Tzadiko commented on src/wallet/fee_algorithm.h:
```diff
@@ -0,0 +1,22 @@
+        HardforkV8,
+    };
+
+    class FeeAlgorithmUtilities
```
I prefer to keep it as a class incase I need to add state to the class later. I'll leave as is since this is a nit.

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+        using EnumStringsType = std::array<std::string_view, 5>;
+        using EnumsType = std::array<FeePriority, 5>;
+
+        static inline const EnumStringsType strings_ = { { "default", "unimportant", "normal", "elevated", "priority" } };
```
Done

Tzadiko commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -164,7 +168,7 @@ static std::string get_human_readable_timespan(uint64_t seconds);
 namespace
 {
-  const std::array<const char* const, 5> allowed_priority_strings = {{"default", "unimportant", "normal", "elevated", "priority"}};
+  const std::array<std::string_view, 5> allowed_priority_strings = FeePriorityUtilities::GetFeePriorityStrings();
```
It's used in 6 places, so I'd rather keep the changeset small by just replacing how it's created.

Tzadiko commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -763,17 +767,13 @@ namespace
 }
 
-bool parse_priority(const std::string& arg, uint32_t& priority)
+bool parse_priority(const std::string& arg, FeePriority& priority)
```
Done

Tzadiko commented:
Writing 'Done' on comments I've addressed locally (to keep track for myself). Will comment here again when I push all the changes.

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+
+        static FeePriority FromIntegral(const uint32_t priority)
+        {
+            return static_cast<FeePriority>(priority);
```
Done

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+
+    std::ostream& operator<<(std::ostream& os, const FeePriority priority);
+
+    class FeePriorityUtilities
```
I'm used to writing utilities as classes. I get that you can have them as free function in namespace scope, and that's valid too. Maybe moreso of a habit to 'sub-group' logical units in classes (i.e. namespace -> class -> method, instead of namespace -> method).

vtnerd replied:
I would say just get use to namespaces, because the remainder of the codebase is written in that style.

Tzadiko replied:
Done

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -8470,42 +8470,44 @@ uint64_t wallet2::estimate_fee(bool use_per_byte_fee, bool use_rct, int n_inputs
   {
-    size_t count;
-    uint64_t multipliers[4];
+    FeePriority maximum_priority; // Determines the maximum priority available for said fee algorithm.
```
What is the scope of 'all of this'? You want me to mark get_fee_multiplier as constexpr? What's the point? Are any of these methods called in a constexpr context?

vtnerd replied:
The values in global `static` memory should be `constexpr` where possible to avoid the initialization order fiasco. This doesn't help us/Monero frequently, as we don't have anything crazy happening before `int main`, but still helps with program correctness.

Tzadiko replied:
I've made fee_steps static constexpr instead of static const.

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+            }
+            else
+            {
+                return priority;
```
The upper bound is clamped by virtue of it being an enum, so not sure what you mean. Do you mean I add another value to fee_priority and call it UpperLimit, so I have

enum class Something
{
    Default
    Unimportant
    ...
    Priority
    Upperbound // Upper clamp
}

I don't think it's a good idea, since it'll be adding more complexity to fee_priority, which already is confusing to understand (we're mapping fee_priority to an actionable fee_priority which is a remnant of how the code was done before.

Tzadiko replied:
Or maybe you mean use std::size on the enum? I'm unsure.

vtnerd replied:
Just some defensive programming. The `enum` value can be cast to any value, and doesn't have to be one of the types defined.

Tzadiko replied:
I still don't understand what you mean here. This method does one think only, that being a lowerbound clamp (Default -> Uniportant). Why would I need to add an upperbound clamp? I understand the need for an upperbound clamp for integer -> FeePriority conversions, but an upperbound clamp doesn't make sense in the context of FeePriority -> FeePriority here.

vtnerd replied:
```c++
fee_priority_utlities::clamp(fee_priority(std::numeric_limits<std::uint32_t>::max()));
```

would `return fee_priority::priority;`

So that `clamp` literally clamps all values to the valid range. Whereas currently it just means "do something if ` == fee_priority::default_`.

Tzadiko replied:
I did not know this (`fee_priority(std::numeric_limits<std::uint32_t>::max())`) was possible. Learning new things. 😃 . I get it now, addressed.

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+        static std::string_view ToString(const FeePriority priority)
+        {
+            const auto integralValue = AsIntegral(priority);
+            return strings_.at(integralValue);
```
Im not sure why I'd need a clamp here. There are five enum values and five strings. They map 1-to-1. I'm assuming you're referring to a case where an enumeration is added later but the `string_` is not changed, in that case there is the opportunity to throw. I'm not sure what a clamp would do there though. I added a comment for the next person to add to the string_ member variable when a fee_priority enum is added.

vtnerd replied:
Same thing here about casting to an `enum`.

Tzadiko replied:
I still don't understand, because the bounding condition can never be exceeded. We're passing in an enum not an integer. There is one string_ for each FeePriority, where each string's index maps to the underlying enum value.

vtnerd replied:
I was suggesting possibly clamping `priority` values to the `0...4` range so that `.at(...)` never throws. Since `.at(std::numeric_limits<std::uin32_t>::max())` has defined behavior its not really needed, and perhaps hides potential bugs.

Tzadiko commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -2474,7 +2482,7 @@ bool simple_wallet::set_default_priority(const std::vector<std::string> &args/*
     if (pwd_container)
     {
-      m_wallet->set_default_priority(priority);
+      m_wallet->set_default_priority(FeePriorityUtilities::FromIntegral(priority));
```
I changed it to use fee_priority enum.

Tzadiko commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -3711,9 +3719,10 @@ bool simple_wallet::set_variable(const std::vector<std::string> &args)
-      priority_string = allowed_priority_strings[priority];
+    const FeePriority priority = m_wallet->get_default_priority();
+    const auto priority_index = FeePriorityUtilities::AsIntegral(priority);
+    if (priority_index < allowed_priority_strings.size())
```
Removed

Tzadiko commented on src/wallet/fee_priority.h:
```diff
@@ -0,0 +1,95 @@
+#include <array>
+#include <algorithm>
+#include <iterator>
+#include <optional>
```
Done

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -4606,7 +4606,7 @@ boost::optional<wallet2::keys_file_data> wallet2::get_keys_file_data(const crypt
   json.AddMember("default_mixin", value2, json.GetAllocator());
 
-  value2.SetUint(m_default_priority);
+  value2.SetUint(static_cast<unsigned>(m_default_priority));
```
Done boost approach.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -5071,15 +5071,15 @@ bool wallet2::load_keys_buf(const std::string& keys_buf, const epee::wipeable_st
     if (field_default_priority_found)
     {
-      m_default_priority = field_default_priority;
+      m_default_priority = static_cast<FeePriority>(field_default_priority);
```
Done boost approach.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -8470,42 +8470,44 @@ uint64_t wallet2::estimate_fee(bool use_per_byte_fee, bool use_rct, int n_inputs
   }
 
-  THROW_WALLET_EXCEPTION_IF(fee_algorithm < 0 || fee_algorithm > 3, error::invalid_priority);
+  THROW_WALLET_EXCEPTION_IF(fee_algorithm == FeeAlgorithm::Unset || fee_algorithm > FeeAlgorithm::HardforkV8, error::invalid_priority);
```
std::size() is a template function and relies on template type deduction. Since it requires the type of the array to be deducible, it won't work if the array is of an anonymous type. Changed the type to a struct (adding the struct fee_step in its place) and made fee_steps an array of type fee_step.

Tzadiko replied:
I don't know what I did to lead me to that conclusion initially. I just tried it with a mock [example](https://godbolt.org/z/W7ndaGa8x) in Godbolt and it worked. I don't know how I feel about reverting it though. What's wrong with being more expressive with a struct (I think you meant 'dummy struct' instead of 'dummy enum')? I think it increases the readability here, which is the primary purpose of this MR.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -5071,15 +5071,15 @@ bool wallet2::load_keys_buf(const std::string& keys_buf, const epee::wipeable_st
       GET_FIELD_FROM_JSON_RETURN_ON_ERROR(json, default_fee_multiplier, unsigned int, Uint, false, 0);
       if (field_default_fee_multiplier_found)
-        m_default_priority = field_default_fee_multiplier;
+        m_default_priority = static_cast<FeePriority>(field_default_fee_multiplier);
```
There is clamping now for the upperbound when going from integer type to enum in FromIntegral.

vtnerd commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -164,7 +167,7 @@ static std::string get_human_readable_timespan(uint64_t seconds);
 namespace
 {
-  const std::array<const char* const, 5> allowed_priority_strings = {{"default", "unimportant", "normal", "elevated", "priority"}};
+  const std::array<std::string_view, 5> allowed_priority_strings = tools::fee_priority_utilities::feePriorityStrings_;
```
Its nitpick, but there's still a bunch of camel case in this patch. Typically everything is snake case.

Tzadiko replied:
See comment below: https://github.com/monero-project/monero/pull/9838#issuecomment-2972073678 Done

vtnerd commented on src/wallet/fee_algorithm.h:
```diff
@@ -0,0 +1,21 @@
+
+    namespace fee_algorithm_utilities
+    {
+        static int AsIntegral(const fee_algorithm algorithm)
```
`as_integral`, etc.

vtnerd reviewed (commented):
The only holdup (as mentioned twice in other comments) is the continued use of camel case. Try to normalize the remainder functions as snake, and this is ready to ship.

Tzadiko commented:
> The only holdup (as mentioned twice in other comments) is the continued use of camel case. Try to normalize the remainder functions as snake, and this is ready to ship.

I don't like it when I need to repeat myself either, so I'm sorry about that. I've implemented the snake_case change everywhere (outside of the value of enumerations). I squashed all commits. Should be good to go now.

vtnerd commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -164,7 +167,7 @@ static std::string get_human_readable_timespan(uint64_t seconds);
 namespace
 {
-  const std::array<const char* const, 5> allowed_priority_strings = {{"default", "unimportant", "normal", "elevated", "priority"}};
+  const std::array<std::string_view, 5> allowed_priority_strings = tools::fee_priority_utilities::fee_priority_strings;
```
Missed this last time. This should be marked `constexpr` as well.

Tzadiko replied:
Thank you, done. Squashed and pushed.

selsta reviewed (approved):
Last force push only fixed whitespace changes.

0xFFFC0000 commented:
> @0xFFFC0000 would we need this for a follow up PR on release branch?

Not at the moment. But since the code makes that part more readable, might be good idea to have it on the list for release too ( since it is already battle-tested ).
